### PR TITLE
Updates for Linux/arm64 compatibility

### PR DESCRIPTION
**Pull request by otterley**
Ensure `other.test_dot_a_all_contents_invalid` and `other.jsmath` pass on arm64. 

Also, make `core.test_freetype` and `core.test_bullet_autoconf` pass on arm64 by updating the (ancient) `config.guess` files in the vendored projects.

Signed-off-by: Michael Fischer <fiscmi@amazon.com>

### CODE REVIEW

File: tests/test_other.py
Comment (sbc100) [inline tests/test_other.py:9352]: I think `assertLess(diff, 5)` is probably simpler here.  WDYT @kripken ?

File: tests/test_other.py
Comment (sbc100) [inline tests/test_other.py:1166]: The `--target` flag can be completely removed once #16561 lands

**Comment by otterley** [comment]
I don't know why test-wasm64 failed - it does not appear to be related to my change. Maybe needs a retry?

File: tests/test_other.py
Comment (otterley) [reply tests/test_other.py:1166]: ack

File: tests/test_other.py
Comment (sbc100) [reply tests/test_other.py:9352]: I think there are two reasons to change this:

1. I'm not a fan of python's if/else ternary expressions (thats a style nit obviously).  I think this makes an already non-obvious line, even less obvious.
2. I think it makes the next needlessly fragile when I think we should be making it less fragile (what happens when another architecture or OS starts to differ here?)

Having said that, if you are too busy to fix, I agree these are both fairly minor reasons.

So.. maybe a 3 out of 10 :)

**Review by sbc100 — APPROVED** [review]

**Comment by sbc100** [comment]
For some reason the test you modified is failing on mac... which is odd because I would expect #16561 to ensure this is still a native object.

**Comment by otterley** [comment]
@sbc100 any good ideas?

**Comment by sbc100** [comment]
It looks like something that would require access to macOS to debug. 

In this test `llvm-ar crs` is supposed to create an archive with an index.. and then emcc is supposed to fail because the object file in that archive is not of the correct kind

For some reason after your change llvm is now reporting that the archive does not have an index at all.

**Comment by sbc100** [comment]
New and unexpected error after this change: `emcc: warning: libfoo.a: archive is missing an index; Use emar when creating libraries to ensure an index is created [-Wemcc]`

**Comment by otterley** [comment]
I have access to Macs. x86 or arm64?

**Comment by sbc100** [comment]
x64 (we don't have any arm machines available on CI)

**Comment by otterley** [comment]
Which MacOS version do you build against?

**Comment by sbc100** [comment]
you should be able to get that info from the failing CI report: https://app.circleci.com/pipelines/github/emscripten-core/emscripten/20114/workflows/9911865d-74f4-4f9c-9236-befe8ceb663a/jobs/518168
Or maybe in the circleci config file  .circleci/config.yml

I think it won't matter though since this tests just uses tools from emsdk (i.e. our own copy of clang / llvm / python etc)

**Comment by otterley** [comment]
Thanks. Looks like XCode 12.2.0 on MacOS 10.15.5 (as of today).

**Comment by otterley** [comment]
Sure enough, this change causes the Mac build to fail the `other.test_dot_a_all_contents_invalid` test.

Old code:

```
% EMTEST_VERBOSE=1 ./tests/runner other.test_dot_a_all_contents_invalid

Prefer --verbose over setting $EMTEST_VERBOSE
Prefer --verbose over setting $EMTEST_VERBOSE
root:DEBUG: Checking JS engine ['/Users/ec2-user/emsdk/node/14.18.2_64bit/bin/node']
Running: '/Users/ec2-user/emsdk/node/14.18.2_64bit/bin/node /Users/ec2-user/emscripten/tests/hello_world.js'
posixtestsuite not found (run git submodule update --init?)
Test suites:
['test_other']
Running test_other: (1 tests)
(checking sanity from test runner)
shared:DEBUG: successfully executed /Users/ec2-user/emsdk/upstream/bin/clang --version
cache:DEBUG: PID 4283 acquiring multiprocess file lock to Emscripten cache at /Users/ec2-user/cache
tools.filelock:DEBUG: Attempting to acquire lock 4316418000 on /Users/ec2-user/cache/cache.lock
tools.filelock:DEBUG: Lock 4316418000 acquired on /Users/ec2-user/cache/cache.lock
cache:DEBUG: done
shared:DEBUG: sanity file up-to-date but check forced: /Users/ec2-user/cache/sanity.txt
shared:DEBUG: successfully executed /Users/ec2-user/emsdk/node/14.18.2_64bit/bin/node --version
shared:DEBUG: successfully executed /Users/ec2-user/emsdk/upstream/bin/llc --version
shared:INFO: (Emscripten: Running sanity checks)
shared:DEBUG: successfully executed /Users/ec2-user/emsdk/node/14.18.2_64bit/bin/node -e console.log("hello")
profiler:DEBUG: block "sanity compiler_engine" took 0.033 seconds
profiler:DEBUG: block "sanity LLVM" took 0.000 seconds
tools.filelock:DEBUG: Attempting to release lock 4316418000 on /Users/ec2-user/cache/cache.lock
tools.filelock:DEBUG: Lock 4316418000 released on /Users/ec2-user/cache/cache.lock
cache:DEBUG: PID 4283 released multiprocess file lock to Emscripten cache at /Users/ec2-user/cache
profiler:DEBUG: block "check_sanity" took 0.061 seconds
test_dot_a_all_contents_invalid (test_other.other) ... shared:DEBUG: successfully executed /Users/ec2-user/emsdk/upstream/bin/clang native.c -target x86_64-linux -c -o native.o
shared:DEBUG: successfully executed /Users/ec2-user/emscripten/emar crs libfoo.a native.o
shared:DEBUG: executed /Users/ec2-user/emscripten/emcc main.c libfoo.a
ok
```

New code (Failing):

```
 % EMTEST_VERBOSE=1 ./tests/runner other.test_dot_a_all_contents_invalid

Prefer --verbose over setting $EMTEST_VERBOSE
Prefer --verbose over setting $EMTEST_VERBOSE
root:DEBUG: Checking JS engine ['/Users/ec2-user/emsdk/node/14.18.2_64bit/bin/node']
Running: '/Users/ec2-user/emsdk/node/14.18.2_64bit/bin/node /Users/ec2-user/emscripten/tests/hello_world.js'
posixtestsuite not found (run git submodule update --init?)
Test suites:
['test_other']
Running test_other: (1 tests)
(checking sanity from test runner)
shared:DEBUG: successfully executed /Users/ec2-user/emsdk/upstream/bin/clang --version
cache:DEBUG: PID 4304 acquiring multiprocess file lock to Emscripten cache at /Users/ec2-user/cache
tools.filelock:DEBUG: Attempting to acquire lock 4351737808 on /Users/ec2-user/cache/cache.lock
tools.filelock:DEBUG: Lock 4351737808 acquired on /Users/ec2-user/cache/cache.lock
cache:DEBUG: done
shared:DEBUG: sanity file up-to-date but check forced: /Users/ec2-user/cache/sanity.txt
shared:DEBUG: successfully executed /Users/ec2-user/emsdk/node/14.18.2_64bit/bin/node --version
shared:DEBUG: successfully executed /Users/ec2-user/emsdk/upstream/bin/llc --version
shared:INFO: (Emscripten: Running sanity checks)
shared:DEBUG: successfully executed /Users/ec2-user/emsdk/node/14.18.2_64bit/bin/node -e console.log("hello")
profiler:DEBUG: block "sanity compiler_engine" took 0.032 seconds
profiler:DEBUG: block "sanity LLVM" took 0.000 seconds
tools.filelock:DEBUG: Attempting to release lock 4351737808 on /Users/ec2-user/cache/cache.lock
tools.filelock:DEBUG: Lock 4351737808 released on /Users/ec2-user/cache/cache.lock
cache:DEBUG: PID 4304 released multiprocess file lock to Emscripten cache at /Users/ec2-user/cache
profiler:DEBUG: block "check_sanity" took 0.061 seconds
test_dot_a_all_contents_invalid (test_other.other) ... shared:DEBUG: successfully executed /Users/ec2-user/emsdk/upstream/bin/clang native.c -c -o native.o --target=x86_64-darwin -isystem /Users/ec2-user/emscripten/system/include/libcxx
shared:DEBUG: successfully executed /Users/ec2-user/emscripten/emar crs libfoo.a native.o
shared:DEBUG: executed /Users/ec2-user/emscripten/emcc main.c libfoo.a
FAIL

======================================================================
FAIL: test_dot_a_all_contents_invalid (test_other.other)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/ec2-user/emscripten/tests/test_other.py", line 1170, in test_dot_a_all_contents_invalid
    self.assertContained('unknown file type', stderr)
  File "/Users/ec2-user/emscripten/tests/common.py", line 817, in assertContained
    self.fail("Expected to find '%s' in '%s', diff:\n\n%s\n%s" % (
AssertionError: Expected to find 'unknown file type
' in 'emcc: warning: libfoo.a: archive is missing an index; Use emar when creating libraries to ensure an index is created [-Wemcc]
emcc: warning: libfoo.a: adding index [-Wemcc]
error: undefined symbol: native (referenced by top-level compiled C/C++ code)
warning: Link with `-s LLD_REPORT_UNDEFINED` to get more information on undefined symbols
warning: To disable errors for undefined symbols use `-s ERROR_ON_UNDEFINED_SYMBOLS=0`
warning: _native may need to be added to EXPORTED_FUNCTIONS if it arrives from a system library
Error: Aborting compilation due to previous errors
emcc: error: '/Users/ec2-user/emsdk/node/14.18.2_64bit/bin/node /Users/ec2-user/emscripten/src/compiler.js /var/folders/4n/xfbzwyvn4dldkk9st7h4vf8h0000gn/T/tmp0dljpxq5.json' failed (returned 1)
', diff:

--- expected
+++ actual
@@ -1 +1,9 @@
-unknown file type
+emcc: warning: libfoo.a: archive is missing an index; Use emar when creating libraries to ensure an index is created [-Wemcc]
+emcc: warning: libfoo.a: adding index [-Wemcc]
+error: undefined symbol: native (referenced by top-level compiled C/C++ code)
+warning: Link with `-s LLD_REPORT_UNDEFINED` to get more information on undefined symbols
+warning: To disable errors for undefined symbols use `-s ERROR_ON_UNDEFINED_SYMBOLS=0`
+warning: _native may need to be added to EXPORTED_FUNCTIONS if it arrives from a system library
+Error: Aborting compilation due to previous errors
+emcc: error: '/Users/ec2-user/emsdk/node/14.18.2_64bit/bin/node /Users/ec2-user/emscripten/src/compiler.js /var/folders/4n/xfbzwyvn4dldkk9st7h4vf8h0000gn/T/tmp0dljpxq5.json' failed (returned 1)
+
```

The key difference is:

`clang native.c -target x86_64-linux -c -o native.o` 
`clang native.c -c -o native.o --target=x86_64-darwin -isystem /Users/ec2-user/emscripten/system/include/libcxx` => Fails

Upon further study, I notice that this test case actually expects a linker failure. It's just that the error message is different now.

Old:

```
wasm-ld: error: unknown file type: libfoo.a(native.o)
```

New:

```
error: undefined symbol: native (referenced by top-level compiled C/C++ code)
```

So, I'm wondering if it makes sense just to allow for both variants of error text. WDYT?

**Comment by sbc100** [comment]
The expectation of this test is that native.o *is* an invalid object file, and should be rejected as a linker input.

It is the purpose of this test that we report a  wrongly compiled object file.   When a user attempts to include an object file that was compiled for the wrong architecture (which is very common with cross compilers) we want to make it as clear as possible that the object file is bad.  

Reported  an "undefined symbol" of much lesser value.   It doesn't include the key information that an invalid object file was passed to the linker.

**Comment by otterley** [comment]
OK. Unfortunately I don't know how to solve this problem except by special-casing it for MacOS to return it to the same code as before, where the target is `x86_64-linux`. Is that the right solution?

**Comment by dschuff** [comment]
OK, I reproduced this on my mac locally. I think this is one of those cases where you investigate an issue like this, only to find that something unrelated has already been broken and you didn't know it :)

In particular it seems that archive indexes are actually never recognized correctly on mac. I think I know why too, but I'll file a separate bug about that. I don't think it needs to block this PR.

File: tests/test_other.py
Comment (dschuff) [inline tests/test_other.py:1167]:
```suggestion
    self.run_process([EMAR, '--format=gnu', 'crs', 'libfoo.a', 'native.o'])
```

**Review by dschuff — COMMENTED** [review]
Can you try adding this? I think that will make the test pass.

File: tests/test_other.py
Comment (otterley) [reply tests/test_other.py:1167]: Unfortunately, that didn't work:

```
shared:DEBUG: successfully executed /Users/ec2-user/emscripten/emar --format=gnu crs libfoo.a native.o
shared:DEBUG: executed /Users/ec2-user/emscripten/emcc main.c libfoo.a
FAIL

======================================================================
FAIL: test_dot_a_all_contents_invalid (test_other.other)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/ec2-user/emscripten/tests/test_other.py", line 1169, in test_dot_a_all_contents_invalid
    self.assertContained('unknown file type', stderr)
  File "/Users/ec2-user/emscripten/tests/common.py", line 817, in assertContained
    self.fail("Expected to find '%s' in '%s', diff:\n\n%s\n%s" % (
AssertionError: Expected to find 'unknown file type
' in 'error: undefined symbol: native (referenced by top-level compiled C/C++ code)
warning: Link with `-s LLD_REPORT_UNDEFINED` to get more information on undefined symbols
warning: To disable errors for undefined symbols use `-s ERROR_ON_UNDEFINED_SYMBOLS=0`
warning: _native may need to be added to EXPORTED_FUNCTIONS if it arrives from a system library
Error: Aborting compilation due to previous errors
emcc: error: '/Users/ec2-user/emsdk/node/14.18.2_64bit/bin/node /Users/ec2-user/emscripten/src/compiler.js /var/folders/4n/xfbzwyvn4dldkk9st7h4vf8h0000gn/T/tmp2how2jnn.json' failed (returned 1)
', diff:

```

**Comment by dschuff** [comment]
I'm going to go ahead and disable this test for now, at least on mac (I'll have to see what happens on Windows). I'll leave the change to use the native clang flags in #16650 so I can test on my M1 mac. I think maybe you should just update this PR to remove all the changes to this test and I can land #16650 first.

**Comment by otterley** [comment]
Any idea why the Firefox test failed?

**Comment by dschuff** [comment]
(sorry, forgot to mention this before, but it looked like flakiness in the local web server that drives the tests. I just restarted it, and it landed).

**Pull request merged**
